### ISSUE TITLE
docs: add continued pretraining (CPT) — new training method from PraisonAI PR #4361

### ISSUE THREAD

**Issue by MervinPraison** [MervinPraison/PraisonAIDocs#3352]
## Summary

[PraisonAI PR #4361](https://github.com/MervinPraison/PraisonAI/pull/4361) (merged 2026-08-25) adds **continued pretraining (CPT)** as a first-class training method in `praisonai-train llm`. It also fixes two blocking bugs that prevented CPT from working at all:

1. **Raw-text corpora now load.** A dataset with only a `text` column used to fall through to the Alpaca branch, produce empty rows, and die with `"All examples formatted to empty text — check dataset schema / chat_template"`. Domain- and language-adaptation runs (the exact use case in this repo's model-training skill) could not run.
2. **Embeddings now get the right learning rate.** `modules_to_save` was exposed, so users could make `embed_tokens`/`lm_head` trainable — but `SFTTrainer` has no `embedding_learning_rate`, so embeddings trained at the (much higher) adapter LR. CPT now uses `UnslothTrainer` / `UnslothTrainingArguments`, which is what Unsloth ships that field on. Default: `learning_rate / 10`.

Nothing about CPT is currently documented in `PraisonAIDocs`. This is both a **new-content** ask (a dedicated CPT feature page) and a **content-update** ask (two edits to `docs/train.mdx` plus one to `docs/cli/train.mdx`).

---

## Source of truth (read before writing)

Per `AGENTS.md` §1.2 "SDK-First Documentation Cycle" — read these first, then document:

- SDK trainer source: [`src/praisonai-train/praisonai_train/train/llm/trainer.py`](https://github.com/MervinPraison/PraisonAI/blob/main/src/praisonai-train/praisonai_train/train/llm/trainer.py) (at merge SHA `8212dd07`)
  - `TRAINING_METHODS["cpt"]` registry entry — trainer `UnslothTrainer`, config `UnslothTrainingArguments`, required columns `("text",)`, no ref model.
  - `formatting_prompts_func` — new raw-text branch when `conversations` and `instruction` are both absent and `text` is present. Passes rows through unchanged; non-string rows become `""`.
  - `TrainModel.process_dataset` — updated guard `self.config.get("method", "sft") not in ("sft", "cpt")`; new empty-dataset error message that names ShareGPT / Alpaca / raw text and prints the columns the dataset actually had.
  - `TrainModel.train_model` — CPT-only block that sets `sft_params["embedding_learning_rate"]` (defaulting to `learning_rate / 10`) and prints a NOTE when `modules_to_save` is unset.
  - `_lazy_import_training_deps()` — pulls `UnslothTrainer`, `UnslothTrainingArguments` (older `unsloth` versions leave them `None`, CPT then reports what to upgrade).
  - Config allowlist now includes `embedding_learning_rate`.
- Tests: [`src/praisonai-train/tests/unit/train/test_continued_pretraining.py`](https://github.com/MervinPraison/PraisonAI/blob/main/src/praisonai-train/tests/unit/train/test_continued_pretraining.py) — 11 tests spelling out the contract.
- PR: https://github.com/MervinPraison/PraisonAI/pull/4361 (`feat/train-continued-pretraining` → `main`, +173/-87, 3 files).

---

### 1. NEW page — `docs/features/praisonai-train-cpt.mdx`

Placement: `docs/features/` per `AGENTS.md` §1.8 (folder placement rules). Do **not** put this in `docs/concepts/`.

Follow the AGENTS.md page-structure template §2:

```
---
title: "Continued Pretraining (CPT)"
sidebarTitle: "Continued Pretraining"
description: "Adapt a base model to a new domain, language, or vocabulary with a raw-text corpus"
icon: "book-open"
---
```

**Hero diagram** (color scheme per §3.1 — `#8B0000` input, `#189AB4` process, `#10B981` output, `#F59E0B` decision, `#6366F1` config, `#fff` text, `#7C90A0` stroke):

Show `📚 raw-text corpus → 🧠 UnslothTrainer (adapter LR + embedding LR) → ✅ adapted model`.

**Sections required**:

- **Quick Start** using `<Steps>` — two steps:
  1. Minimal: `method: cpt` + a dataset with a `text` column + `modules_to_save: [embed_tokens, lm_head]`.
  2. With a custom `embedding_learning_rate` (e.g. `2e-5`).
  
  Every example must be agent-centric per AGENTS.md §1.1 rule 9 — start with the config a user actually writes, not with library internals.

- **How It Works** — one sequence diagram showing `User → config.yaml → praisonai-train llm → UnslothTrainer → adapted model`. Then a short table explaining why CPT is separate from SFT (needs a lower embedding LR; needs `modules_to_save` to move the vocab).

- **When to use CPT** — a small mermaid decision diagram per AGENTS.md §6.1 ("If multiple options in one page, people might be confused on what to choose, so create the mermaid diagram to choose what option at what instance"). Nodes: "adapt to a new language?", "adapt to a new domain?", "teach specific instructions?" → SFT / DPO / ORPO / KTO / CPT.

- **Dataset shape** — CPT needs a dataset with a `text` column. ShareGPT (`conversations`) and Alpaca (`instruction`/`input`/`output`) still take priority when present; a stray `text` column on those shapes is ignored. Non-string rows become `""` and are then dropped. Quote the new empty-dataset error verbatim so users can search for it: `"All examples formatted to empty text. The dataset columns are [...]; this trainer understands ShareGPT ('conversations'), Alpaca ('instruction'/'input'/'output') and raw text ('text'). Rename your columns to one of those shapes, or set chat_template to match your data."`

- **Configuration** — table of the two CPT-specific keys:

  | Key | Type | Default | Description |
  |---|---|---|---|
  | `method` | `str` | `"sft"` | Set to `"cpt"` to enable continued pretraining. |
  | `embedding_learning_rate` | `float` | `learning_rate / 10` | Learning rate for `embed_tokens` and `lm_head` when they are in `modules_to_save`. |
  | `modules_to_save` | `list[str]` | *unset* | Include `embed_tokens` and `lm_head` to actually move the vocab. Without this, CPT trains no embeddings and prints a NOTE. |

- **Common Patterns** — two:
  1. **Domain adaptation** (e.g. medical / legal corpus): `method: cpt` + `modules_to_save: [embed_tokens, lm_head]` on the model's own tokenizer.
  2. **Language adaptation** (e.g. Tamil): same, plus a tokenizer whose vocab covers the target script.

- **Best Practices** in `<AccordionGroup>`:
  - Always set `modules_to_save: [embed_tokens, lm_head]` — CPT without it trains no embeddings.
  - Keep `embedding_learning_rate` at least 10× below `learning_rate` (the default). The known-bad recipe is training embeddings at the adapter LR — that is why `UnslothTrainer` exists.
  - Upgrade `unsloth` if `UnslothTrainer` / `UnslothTrainingArguments` are unavailable (older `unsloth` versions leave them `None` and CPT will tell you which package to upgrade).
  - After CPT, follow up with SFT for task adaptation — CPT teaches the model the corpus; SFT teaches it to follow instructions on it.

- **Related** `<CardGroup cols={2}>` — link to `/docs/train`, `/cli/train`, and `/features/praisonai-train-multigpu`.

**Example config to include in the page** (the PR's own usage block, verbatim so users can copy-paste):

```yaml
method: cpt
modules_to_save: [embed_tokens, lm_head]
embedding_learning_rate: 2e-5      # defaults to learning_rate / 10
```

---

### 2. UPDATE — `docs/train.mdx`

- **Config keys table (LLM fine-tuning)** — add a new row for `embedding_learning_rate` (`float`, default `learning_rate / 10`, "Only used when `method: cpt`. Rate for `embed_tokens`/`lm_head`; the default is the recipe `UnslothTrainer` was written to enforce.")
- **`modules_to_save` row (line ~933)** — add a follow-up sentence: "Required for `method: cpt` to actually adapt the vocabulary. Without it, CPT trains no embeddings and prints a NOTE."
- **New short subsection under "Fine-tuning Setup"** titled **Training method (`method:`)** — a one-line explainer that lists the five supported methods with a table:

  | `method:` | Trainer used | Required dataset columns | Notes |
  |---|---|---|---|
  | `sft` (default) | `SFTTrainer` | ShareGPT or Alpaca (formatted to `text`) | Standard supervised fine-tuning. |
  | `cpt` | `UnslothTrainer` | `text` | Continued pretraining — raw-text corpus. See [Continued Pretraining](/features/praisonai-train-cpt). |
  | `dpo` | `DPOTrainer` | `prompt`, `chosen`, `rejected` | Preference. |
  | `orpo` | `ORPOTrainer` | `prompt`, `chosen`, `rejected` | Preference. |
  | `kto` | `KTOTrainer` | `prompt`, `completion`, `label` | Preference. |
- **Empty-dataset error snippet / troubleshooting** — if there is a troubleshooting or "common errors" section, add the new message and its three-shape hint so users searching for it land here.

### 3. UPDATE — `docs/cli/train.mdx`

- In the **Config keys for LLM fine-tuning** table, add rows for `method` (default `"sft"`, "One of `sft`, `cpt`, `dpo`, `orpo`, `kto`") and `embedding_learning_rate` (see above).
- In the **Supported models** / examples area, add a one-line note that `method: cpt` with `modules_to_save: [embed_tokens, lm_head]` enables continued pretraining on a raw-text corpus, and link to the new `/features/praisonai-train-cpt` page.

### 4. UPDATE — `docs/docs.json`

Add `"features/praisonai-train-cpt"` to the Features group next to the other `features/praisonai-train-*` entries (around line 1100 — see `praisonai-train-package`, `praisonai-train-multigpu`, `praisonai-train-checkpointing`, `praisonai-train-dataset-tooling`, `praisonai-train-benchmark`, `praisonai-train-export`, `praisonai-train-serve`). Validate JSON after editing.

---

## Constraints from AGENTS.md the follow-up agent must honour

- Read the SDK source first — never guess signatures (§1.9 rule 3).
- Do **not** create or modify files in `docs/concepts/` — human-approved only (§1.8).
- Use Mintlify components (`<Steps>`, `<AccordionGroup>`, `<CardGroup>`) and a hero mermaid diagram with the standard color scheme (§1.9 rules 9, 10).
- Every code example must run without modification and include all imports (§5.1).
- One-sentence section intros, active voice, no forbidden phrases like "In this section, we will…" (§6).
- Friendly imports — the trainer is driven from `config.yaml`, not from a `praisonaiagents` import, so all examples should be YAML.
- Frontmatter must include `title`, `sidebarTitle`, `description`, `icon` (§8.1). Suggested icon: `book-open` or `graduation-cap` — a Lucide name.
- Non-developer focused: after reading, a user should feel "is it really this easy? — just add `method: cpt` and `modules_to_save`?"

---

## References

- PR: https://github.com/MervinPraison/PraisonAI/pull/4361
- Head SHA: `8212dd07eaaac75f890e4a7f904d77b56014b99f`
- Files touched:
  - `src/praisonai-train/praisonai_train/train/llm/trainer.py` (+50, -3)
  - `src/praisonai-train/tests/unit/train/test_continued_pretraining.py` (+119, new)
  - `src/praisonai-train/tests/unit/train/test_training_methods.py` (+4, -84)
- Upstream Unsloth reference cited in the code: `unsloth/trainer.py:445-524` (`UnslothTrainer` is where `embedding_learning_rate` lives).
